File: zero-server/src/main/kotlin/com/zeroday/security/RateLimiter.kt
package com.zeroday.security

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/**
 * Simple in-memory sliding-window rate limiter. Suitable for a single-node
 * deployment; for a clustered server swap [Storage] for a Redis-backed
 * implementation without touching call sites.
 */
class RateLimiter(
    private val maxRequests: Int,
    private val windowMs: Long,
    private val clock: () -> Long = System::currentTimeMillis
) {
    interface Storage {
        fun hitAndCount(key: String, now: Long, windowMs: Long): Int
        fun reset(key: String)
    }

    private val storage = object : Storage {
        private val buckets = ConcurrentHashMap<String, Bucket>()

        override fun hitAndCount(key: String, now: Long, windowMs: Long): Int {
            val bucket = buckets.compute(key) { _, current ->
                if (current == null || now - current.windowStart >= windowMs) {
                    Bucket(now, AtomicLong(1))
                } else {
                    current.count.incrementAndGet()
                    current
                }
            }!!
            return bucket.count.get().toInt()
        }

        override fun reset(key: String) {
            buckets.remove(key)
        }
    }

    fun tryAcquire(key: String): Boolean {
        val count = storage.hitAndCount(key, clock(), windowMs)
        return count <= maxRequests
    }

    /** Reset the rate-limit counter for [key], allowing an immediate retry. */
    fun reset(key: String) = storage.reset(key)

    private data class Bucket(val windowStart: Long, val count: AtomicLong)
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Security/RateLimiter.cs
using System.Collections.Concurrent;

namespace ZeroDayMMO.Server.Security;

public interface IRateLimiterStorage
{
    int HitAndCount(string key, long now, long windowMs);
    void Reset(string key);
}

public class RateLimiter
{
    private readonly IRateLimiterStorage _storage;

    public RateLimiter() : this(new InMemoryRateLimiterStorage()) { }

    public RateLimiter(IRateLimiterStorage storage)
    {
        _storage = storage;
    }

    public bool TryConsume(string key, int limit, TimeSpan window)
    {
        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        var count = _storage.HitAndCount(key, now, (long)window.TotalMilliseconds);
        return count <= limit;
    }

    public void Reset(string key) => _storage.Reset(key);
}

public class InMemoryRateLimiterStorage : IRateLimiterStorage
{
    private readonly ConcurrentDictionary<string, Bucket> _buckets = new();

    public int HitAndCount(string key, long now, long windowMs)
    {
        var bucket = _buckets.AddOrUpdate(key,
            _ => new Bucket { WindowStart = now, Count = 1 },
            (_, current) =>
            {
                if (now - current.WindowStart >= windowMs)
                    return new Bucket { WindowStart = now, Count = 1 };
                var count = Interlocked.Increment(ref current.Count);
                return current;
            });
        return bucket.Count;
    }

    public void Reset(string key)
    {
        _buckets.TryRemove(key, out _);
    }

    private sealed class Bucket
    {
        internal long WindowStart;
        internal int Count;
    }
}
